### ISSUE TITLE
FR: Support nested objects / embedding objects

### ISSUE THREAD

**Issue by samos123** [weaviate/weaviate#2424]
Currently you have to use crossReferences but for some use cases it might be beneficial to have a nested object. For example, right now we would need to create a class for each nested object in order to support importing data from MongoDB with nested objects.

Example data structure:
```
Customer {
name: string
addresses: []Object
[{
   street: string,
   city: string
},]
size: Object
{
   width: int
   height: int
   metric: string
}
```
Note both support a list of objects and support a field of type object is important.

Context: In the Airbyte Connector for Weaviate right now I'm planning to convert nested objects to a json string since creating a new class for each nested object dynamically and creating the reference is quite cumbersome.

Open questions:
1. Should we support making the nested object part of the vector?
2. Should we support only having parts of the nested object be part of the vector?

**Comment by johnballentine** [comment]
This would be huge for us.  We spend so much time figuring out cross-references to represent our nested datasets.

**Comment by dandv** [comment]
If nesting were implemented, that would nicely solve an [FAQ we received](https://weaviate.io/developers/weaviate/current/more-resources/faq.html#q-are-there-any-best-practices-or-guidelines-to-consider-when-designing-a-schema-eg-if-i-was-looking-to-perform-a-semantic-search-over-a-the-content-of-a-book-would-i-look-to-have-chapter-and-paragraph-represented-in-the-schema-etc-would-this-be-preferred-over-including-the-entire-content-of-the-novel-in-a-single-property) about how to represent a `Book` object: `Chapter`s vs. `Paragraph`s vs. sentences.

**Comment by adaptivedavid** [comment]
Thanks for picking this one up. I appreciate that #2477 (Matching Filters on Single Objects) is going to be separate from this, since #2477 explicitly deals with cross-references. That said, will you be adding a similar ability to match nested objects on multiple criteria? For instance, for the sample Customer data structure @samos123 set out, will there be the ability to find all Customers that have an `addresses` item with a `street` of "ABC Street" *and* a `city` of "Big City"?

**Comment by bobvanluijt** [comment]
I've tested the [latest preview](https://hub.docker.com/layers/semitechnologies/weaviate/preview-implement-grpc-nested-objects-c6263d0/images/sha256-63accec2f2705c74702d3fdc881896b8766d1d1f6cbe1fa4873f52416cd2a21d?context=explore) version and noticed an issue with different types inside arrays.

E.g.,

The following examples result into an error.

```json
{
    "q1": [
        "a",
        true
    ]
}
```

```
weaviate.exceptions.UnexpectedStatusCodeException: Creating object! Unexpected status code: 422, with response body: {'error': [{'message': "invalid object: invalid text array property 'q1' on class 'Test0001': invalid text array value: [a %!s(bool=true)]"}]}.
```

---

```json
{
    "q1": {
        "q2": [
            [
                {
                    "a": "b"
                },
                {
                    "c": "d"
                }
            ]
        ]
    }
}
```

```
weaviate.exceptions.UnexpectedStatusCodeException: Creating object! Unexpected status code: 422, with response body: {'error': [{'message': "invalid object: invalid object property 'q1' on class 'Test0002': property 'q1.q2': invalid text property 'q1.q2' on class 'Test0002': not a string, but []interface {}"}]}.
```

**Comment by etiennedi** [comment]
Good catch. The question is: Do we want to support this? It would be valid JSON, but it would also make filtering impossible for that specific datatype, since we need to decide on exactly one type to be able to build a filter-index. (E.g. in the above, we either need to index as `bool` or as `string`). 

How do other DBs deal with this?

**Comment by aliszka** [comment]
I can confirm issues with case (2). Will look into that.
As for case (1) - it is not supported at the moment. Nested objects use same data types as regular properties (except `phone`, `geo` and `refs`). It would accept array of strings, or array of bools, but not mix of types.

**Comment by etiennedi** [comment]
Initial version released in `v1.22`. Future improvements tracked [here](https://github.com/weaviate/weaviate/issues/3694).
